File: 142.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if not head or not head.next: return None
        slow, fast = head, head
        while fast and fast.next:
            if fast is slow:
                slow = head
                while slow is not fast:
                    slow = slow.next
                    fast = fast.next
                return slow
            slow = slow.next
            fast = fast.next.next
        return None

File: 142.go
package leetcode

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}
	slow, fast := head.Next, head.Next.Next
	for slow != fast {
		slow = slow.Next
		if fast == nil || fast.Next == nil {
			return nil
		}
		fast = fast.Next.Next
	}
	// fmt.Println(slow.Val, fast.Val)
	for head != fast {
		head = head.Next
		fast = fast.Next
	}

	return head
}
